File: examples/cpp/minimal-order-service/tests/test_domain.cpp
#include <cassert>
#include <iostream>
#include <memory>
#include <stdexcept>
#include "domain/money.hpp"
#include "domain/order.hpp"

void test_money_creation() {
    domain::Money money(10000, "USD");
    assert(money.amount() == 10000);
    assert(money.currency() == "USD");
    std::cout << "✓ test_money_creation passed\n";
}

void test_money_negative_amount() {
    try {
        domain::Money money(-100, "USD");
        assert(false && "Should have thrown exception");
    } catch (const std::invalid_argument&) {
        std::cout << "✓ test_money_negative_amount passed\n";
    }
}

void test_money_add() {
    domain::Money money1(10000, "USD");
    domain::Money money2(5000, "USD");
    domain::Money result = money1.add(money2);
    assert(result.amount() == 15000);
    assert(result.currency() == "USD");
    std::cout << "✓ test_money_add passed\n";
}

void test_order_creation() {
    auto total = std::make_shared<domain::Money>(10000, "USD");
    auto order = domain::Order::create(total);
    assert(!order->id().empty());
    assert(order->total()->amount() == 10000);
    assert(order->status() == domain::OrderStatus::PENDING);
    std::cout << "✓ test_order_creation passed\n";
}

void test_order_mark_as_paid() {
    auto total = std::make_shared<domain::Money>(10000, "USD");
    auto order = domain::Order::create(total);
    order->markAsPaid();
    assert(order->status() == domain::OrderStatus::PAID);
    std::cout << "✓ test_order_mark_as_paid passed\n";
}

int main() {
    std::cout << "Running domain tests...\n";
    test_money_creation();
    test_money_negative_amount();
    test_money_add();
    test_order_creation();
    test_order_mark_as_paid();
    std::cout << "All domain tests passed!\n";
    return 0;
}

File: examples/python/minimal-order-service/tests/test_domain.py
"""Tests for domain layer."""
import pytest
from domain.money import Money
from domain.order import Order, OrderStatus


def test_money_creation():
    """Test creating money with valid values."""
    money = Money(10000, "USD")
    assert money.amount == 10000
    assert money.currency == "USD"


def test_money_negative_amount():
    """Test that negative amount raises ValueError."""
    with pytest.raises(ValueError, match="Amount cannot be negative"):
        Money(-100, "USD")


def test_money_empty_currency():
    """Test that empty currency raises ValueError."""
    with pytest.raises(ValueError, match="Currency cannot be empty"):
        Money(100, "")


def test_money_add():
    """Test adding money with same currency."""
    money1 = Money(10000, "USD")
    money2 = Money(5000, "USD")
    result = money1.add(money2)
    assert result.amount == 15000
    assert result.currency == "USD"


def test_money_add_different_currencies():
    """Test that adding different currencies raises ValueError."""
    money1 = Money(10000, "USD")
    money2 = Money(5000, "EUR")
    with pytest.raises(ValueError, match="Cannot add different currencies"):
        money1.add(money2)


def test_order_creation():
    """Test creating an order."""
    total = Money(10000, "USD")
    order = Order.create(total)
    assert order.id is not None
    assert order.total == total
    assert order.status == OrderStatus.PENDING


def test_order_mark_as_paid():
    """Test marking an order as paid."""
    total = Money(10000, "USD")
    order = Order.create(total)
    order.mark_as_paid()
    assert order.status == OrderStatus.PAID


def test_order_mark_as_paid_twice():
    """Test that marking a non-pending order as paid raises ValueError."""
    total = Money(10000, "USD")
    order = Order.create(total)
    order.mark_as_paid()
    with pytest.raises(ValueError, match="Only pending orders can be marked as paid"):
        order.mark_as_paid()
